doc/reference/index.rst

```diff
@@ -23,3 +23,268 @@ API Reference
    usb/index.rst
    usermode/index.rst
    runtime_conf/index.rst
+
+Terminology
+***********
+
+The following terms may be used as shorthand API tags to indicate the
+allowed calling context (thread, ISR, pre-kernel), the effect of a call
+on the current thread state, and other behavioral characteristics.
+
+:ref:`api_term_reschedule`
+   if executing the function reaches a reschedule point
+:ref:`api_term_sleep`
+   if executing the function can cause the invoking thread to sleep
+:ref:`api_term_no-wait`
+   if a parameter to the function can prevent the invoking thread from
+   trying to sleep
+:ref:`api_term_isr-ok`
+   if the function can always be safely called from interrupt context
+   even if it may return an error in that case
+:ref:`api_term_pre-kernel-ok`
+   if the function can be safely called before the kernel has been fully
+   initialized, even if it may return an error in that case
+:ref:`api_term_async`
+   if the function may return before the operation it initializes is
+   complete (i.e. function return and operation completion are
+   asynchronous)
+:ref:`api_term_supervisor`
+   if the calling thread must have supervisor privileges to execute the
+   function
+
+Details on the behavioral impact of each attribute are in the following
+sections.  Attributes for some common kernel API are below:
+
+.. list-table::
+   :header-rows: 1
+
+   * - name
+     - reschedule
+     - sleep
+     - no-wait
+     - isr-ok
+     - async
+     - supervisor
+   * - :cpp:func:`k_busy_wait`
+     - no
+     - no*
+     - no*
+     - yes*
+     - no
+     - no
+   * - :cpp:func:`k_sem_give`
+     - yes
+     - no
+     - no*
+     - yes*
+     - no
+     - no
+   * - :cpp:func:`k_sleep`
+     - yes*
+     - yes
+     - no
+     - no
+     - no
+     - no
+   * - :cpp:func:`k_sem_take`
+     - yes*
+     - yes
+     - yes
+     - no
+     - no
+     - no
+   * - :cpp:func:`k_thread_create`
+     - ?
+     - ?
+     - ?
+     - ?
+     - no
+     - yes
```

Comment:
this is wrong, k_thread_create can be called from user mode (subject to some restrictions)
Why did you think it can't? The prototype has `__syscall` in it

Reply:
Also I really hate that we are duplicating information. This stuff should go in the documentation for the APIs themselves and not maintained here.

Reply:
I picked `k_thread_create` due to an off-by-one error associating kernel API functions with attributes from #21389.  Replaced by `k_thread_access_grant`.

We are absolutely not documenting the API in this location.  This is only to get the terms defined so it can be done in the headers where it belongs.  The table is meant as an example; I've revised the wording so that's more clear.

Reply:
Except k_thread_access_grant isn't supervisor only either.

> We are absolutely not documenting the API in this location. 

Yes you are. You are making a table with several kernel APIs and their properties, duplicating information that will be in the documentation of the APIs themselves. These will need to be kept in sync. NACK.

Reply:
There's a simple solution to both problems: no examples in the documentation.  Done.

We can start using these attributes for new API once they're documented.  People who have the expertise to specify the correct attributes for each function can back-fill existing API as needed.

Reply:
> > We are absolutely not documenting the API in this location.
> 
> Yes you are. You are making a table with several kernel APIs and their properties, duplicating information that will be in the documentation of the APIs themselves. These will need to be kept in sync. NACK.

FWIW I found it very useful to have a table of examples next to the definitions themselves. I think the resulting document is less readable and harder to understand without it.